### PR TITLE
Remove a few redundant diagonal `ldiv!` methods

### PR DESCRIPTION
**Pull request by mcognetta**
Three `ldiv!` methods for diagonal matrices were effectively the exact same but had different type specializations. This PR replaces them with a single `ldiv!` that subsumes their functionality.

### CODE REVIEW

File: stdlib/LinearAlgebra/src/diagonal.jl [hunk 1]
```diff
@@ -448,6 +435,7 @@ function ldiv!(D::Diagonal{T}, V::AbstractMatrix{T}) where {T}
     end
     V
 end
+
```
Comment (KlausC) [inline stdlib/LinearAlgebra/src/diagonal.jl:438]: Compared to the code below, the order of the i- and j-loops is different. That means, the operations on `B` are performed row-wise, which is worse than column-wise in most cases. Performance checks with wide `B` required.

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/diagonal.jl:438]: I agree. We should make the `StridedVecOrMat` version be the `AbstractVecOrMat` versionn.

File: stdlib/LinearAlgebra/src/diagonal.jl [hunk 2]
```diff
@@ -419,20 +419,7 @@ mul!(C::AbstractMatrix, A::Transpose{<:Any,<:Diagonal}, B::Transpose{<:Any,<:Rea
...
-        d = D.diag[i]
-        if iszero(d)
-            throw(SingularException(i))
-        end
-        v[i] = d\v[i]
-    end
-    v
-end
-function ldiv!(D::Diagonal{T}, V::AbstractMatrix{T}) where {T}
+function ldiv!(D::Diagonal, V::AbstractVecOrMat)
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/diagonal.jl:422]: ... and let's remove this one also.

File: stdlib/LinearAlgebra/src/diagonal.jl [hunk 3]
```diff
@@ -610,24 +598,6 @@ for f in (:exp, :log, :sqrt,
...
-        for i = 1:m
-            di = D.diag[i]
-            if di == 0
-                throw(SingularException(i))
-            end
-            B[i,j] = di \ B[i,j]
-        end
-    end
-    return B
-end
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/diagonal.jl:630]: Let's keep this, but make it `B::AbstractVecOrMat`.

**Review by dkarrasch — DISMISSED** [review]
Two comments (summarizing what @KlausC said), and then this is ready to go.

**Comment by dkarrasch** [comment]
Hm, actually, this really requires benchmarking. The strided version is optimized towards good access pattern for `B`, whereas the abstract version is optimized towards less reading of the diagonal elements of `D`. I think before changing anything here, we need to see some benchmarks.

**Comment by mcognetta** [comment]
@dkarrasch I used the changes you and @KlausC suggested. Benchmarking revealed only minor performance differences:

Julia 1.5.0

```julia
julia> D = Diagonal(rand(500)); M = rand(500, 500); @btime ldiv!(D, M)
  580.819 μs (0 allocations: 0 bytes)

julia> D = Diagonal(rand(500)); M = UpperTriangular(rand(500, 500)); @btime ldiv!(D, M)
  533.071 μs (0 allocations: 0 bytes)
```


This PR:

```julia
julia> D = Diagonal(rand(500)); M = rand(500, 500); @btime ldiv!(D, M)
  536.923 μs (0 allocations: 0 bytes)

julia> D = Diagonal(rand(500)); M = UpperTriangular(rand(500, 500)); @btime ldiv!(D, M)
  521.033 μs (0 allocations: 0 bytes)
```

File: stdlib/LinearAlgebra/src/diagonal.jl (hunk 2 above)
Comment (mcognetta) [reply stdlib/LinearAlgebra/src/diagonal.jl]: Done, see thread below. Thanks for the comments.

**Review by dkarrasch — COMMENTED** [review]
The benchmark suggests that it makes sense to switch to the previous strided matrix loop order also for abstract matrices. This LGTM. The only question I have is whether we should try to preserve structure when the matrix has structure.

**Pull request merged**